Add ability to use pybids db

kaitj opened the pull request:
This PR adds in functionality to save the `pybids` db. 

It adds a `pybids_db` to the config, where it stores the db directory, and a flag as to whether the db should be overwritten. Wanted this to be separate from the `pybids_inputs`. This also updates the `generate_input.py` to make use of these config options. If the directory is left blank (e.g. `''`), the database will be set to `None` (if left as `''`, it saves the db in the current working directory). Also added in a few small test cases to make sure the db is working as expected.

~@pvandyken I didn't end up using the `pytest.fixtures` and `tmpdir` (mentioned last week) to create the new temporary data during testing - looked as if these temp files would be created at the beginning of the test which would have been picked up by `pybids` and wanted to test for changes.~

pvandyken commented on snakebids/project_template/{{cookiecutter.app_name}}/config/snakebids.yml:
```diff
@@ -33,11 +33,16 @@ pybids_inputs:
+# by default, database is not reindexed
+pybids_db:
+  database_dir: '/path/to/db_dir' # Leave blank if you do not wish to use this
+  write_database: False
```
I find this somewhat confusing: on first glance it looks as if `write_database == False` precludes any database from being written, even if `database_dir` is set. This is not the case: `write_database == False` only prevents a reset of the database. If not database exists, one will be created if a valid path is given.

I would resolve this one of two ways:
1. Keep existing behaviour, but rename `write_database` to `reset_database`. Comment out all three of these lines by default in order to make database use explicitly opt-in.
2. Modify the behaviour so that `write_database == False` actually does prevent any database from being written.

kaitj replied:
Will change to `reset_database` - end of the day the writing is handled by `pybids`

kaitj replied:
This is now changed to `reset_database`.

pvandyken commented on snakebids/tests/test_generate_inputs.py:
```diff
@@ -197,3 +207,51 @@ def test_get_lists_from_bids():
+
+def test_db():
+    # create config
+    real_bids_dir = "snakebids/tests/data/bids_t1w"
```
This test will fail if it gets interrupted and run again, as `{real_bids_dir}/sub-003/anat` won't be deleted.

I'd use the pytest fixture `tmpdir`. Start by using `shutil.copytree` to copy the entire `{real_bids_dir}` into `tmpdir`. Then proceed with the rest of the test as written, using the newly created temporary dataset. That way, the original, source-controlled test folders are never touched. Pytest will take care of cleanup (by default it saves files from the last two tests), so `shutil.rmtree` would no longer necessary.

kaitj replied:
I'm open to it, though when I tried to use the pytest fixture, the folders would be created at the start. I guess the way around it is to create a temp directory, copy the existing files over, and the make the new subject. Will look into it later this week.

kaitj replied:
Took a bit of finagling, but got this to working with `pytest.fixture`. It should be creating testing and updating in a `tmpdir` now.

pvandyken commented on snakebids/tests/test_generate_inputs.py:
```diff
@@ -197,3 +207,51 @@ def test_get_lists_from_bids():
+    assert not pybids_db.get("database_dir")
+
+    # Test saving of new layout (update config)
+    pybids_db["database_dir"] = "snakebids/tests/data/.db/"
```
The db should be saved to a temporary directory `tmpdir/".db"`. This is to prevent a failing test if a previous run is interrupted.

pvandyken commented on snakebids/project_template/{{cookiecutter.app_name}}/workflow/Snakefile:
```diff
@@ -10,6 +10,7 @@ config.update(
     snakebids.generate_inputs(
         bids_dir=config["bids_dir"],
         pybids_inputs=config["pybids_inputs"],
+        pybids_db=config["pybids_db"],
```
The `pybids_db` part of config should not, in my opinion, be strictly necessary. If that is so, this line should be `config.get("pybids_db")`

kaitj replied:
Updated

kaitj commented on snakebids/project_template/{{cookiecutter.app_name}}/config/snakebids.yml:
```diff
@@ -33,11 +33,16 @@ pybids_inputs:
+#this configures the options to save the BIDSLayout
+# by default, database is not reindexed
+pybids_db:
+  database_dir: '/path/to/db_dir' # Leave blank if you do not wish to use this
```
This was just following in the input / output paths in the template. We can set it to where we would like, but we should decide on where that would be. Also I can't recall, but if the default behavior we want is to not make use of this then it should actually be set to `''`

pvandyken replied:
Yeah, the actual file name doesn't matter too much.

I agree the default behaviour should be to have the db disabled, but I also think it would be useful to have `database_dir` prepopulated with a good path choice to set an informal standard for where the database should live. I think a good compromise here would be to comment out all three lines here, a practice commonly used in config files to showcase relevant settings. An explanatory comment can tell users to uncomment the pybids_db block if they wish to use the database

kaitj replied:
Yes, I like this - I've commented it out locally but will wait to push changes until we have decided on an example database path to include

kaitj commented on snakebids/core/input_generation.py:
```diff
@@ -15,6 +15,7 @@
 def generate_inputs(
     bids_dir,
     pybids_inputs,
+    pybids_db=None,
```
Open to splitting it - thought here was to db parameters together similar to how the bids inputs are together.

pvandyken replied:
I'm fine with grouping them in the config, but I don't think they should be glued together in the function signature. The difference between the db params and bids inputs is that bids inputs can have any arbitrary keys which, in turn, contain the data. The db param structure is fixed and static, thus well suited to have explicit params in `generate_inputs`

kaitj commented on snakebids/tests/test_generate_inputs.py:
```diff
@@ -18,11 +23,13 @@ def test_t1w():
     with pytest.raises(ValueError) as v_error:
         config = generate_inputs(
             pybids_inputs=pybids_inputs,
+            pybids_db=pybids_db,
```
Ended up just reverting this back to what it was originally, and `pybids_db` is only called in the `test_db()`

kaitj commented on snakebids/tests/test_generate_inputs.py:
```diff
@@ -197,3 +207,51 @@ def test_get_lists_from_bids():
+    pybids_db = {"database_dir": "", "write_database": False}
+
+    # Test non-saving first
+    _ = _gen_bids_layout(bids_dir=real_bids_dir, derivatives=False, pybids_db=pybids_db)
```
Fixed in latest commits

kaitj commented on snakebids/tests/test_generate_inputs.py:
```diff
@@ -197,3 +202,40 @@ def test_get_lists_from_bids():
+    # Test reading of old layout when changes occur
+    os.makedirs(f"{tmpdir}/data/sub-003/anat")
+    # tmpdir.mkdir("data/sub-003")
+    # tmpdir.mkdir("data/sub-003/anat")
```
Oh, forgot to remove them, was testing out the difference between them - `tmpdir` only lets you a subdirectory a single level down.

kaitj commented:
Things to add / change:
* For relative pathing, relative to `bids_dir`
* sqlite file sould be added to a `.bidsignore`(user handled) for validation purposes -> print out warning message
* separate out `pybids_db` in the function

kaitj commented:
Most of the things above are now fixed. 

I didn't end up including a warning message for the relative path case - need to think about this one a bit more as it would depend on where that db is stored (e.g. in the input directory itself vs a derivatives folder down the line).

Also ended up just adding a comment about the relative pathing in the boilerplate for now. Functionally, this should be working. I haven't been able to test it on a real use case, and only been testing on the test cases that have been implemented.

tkkuehn commented on snakebids/tests/test_generate_inputs.py:
```diff
@@ -197,3 +202,58 @@ def test_get_lists_from_bids():
+        pybids_database_dir=pybids_db.get("database_dir"),
+        pybids_reset_database=pybids_db.get("reset_database"),
+    )
+    assert not pybids_db.get("database_dir")
```
I'm not sure this assertion achieves anything (even if the DB were made on the filesystem, I don't think this would catch it). Not sure though! An alternative might be to check that everything in `{tmpdir}/data` was in the original test data directory.

kaitj replied:
Hmm, I know it will throw an error if the assertion was `assert pybids_db.get('database_dir')`, though I do agree thinking about this a bit more that this may not be the best check. I like the idea of checking against the original data directory.

kaitj replied:
I kept this check in and added another that compares the two directories after it is copied over.

tkkuehn reviewed (changes requested):
A couple of nitpicks but this overall looks good to me!

As far as the (default) location, having thought about it a bit more, I think it makes sense to put it somewhere in the derivatives/output folder, just because it would be good to avoid making any changes to the input directory generally.

kaitj commented on snakebids/project_template/{{cookiecutter.app_name}}/config/snakebids.yml:
```diff
@@ -33,41 +33,47 @@ pybids_inputs:
       - run
-   
+
+#this configures the options to save the BIDSLayout
```
Fair, this may be something worth doing in another PR - just to go through an fix any spacing issues since it seems to be mixed in a number of files. I tried to keep it consistent as best as I could, but I agree, I prefer the space after the `#` too.

kaitj commented on snakebids/core/input_generation.py:
```diff
@@ -233,6 +229,40 @@ def generate_inputs(
+    """Create (or reindex) the BIDSLayout if one doesn't exist,
+    which is only saved if a database directory path is provided
+    """
+    if os.path.exists(bids_dir):
```
> A couple of nitpicks but this overall looks good to me!
> 
> As far as the (default) location, having thought about it a bit more, I think it makes sense to put it somewhere in the derivatives/output folder, just because it would be good to avoid making any changes to the input directory generally.

I've been going in circles with the default location. We would either have to assume the outputs will follow the BIDS structure (e.g. `bids_dir/derivatives`) or know where the derivatives/output folder lives (e.g. in the case of some of the folders we have where `bids` and `deriv` are separate). 

At one point I did think it might be better for it to live in the derivativse/output folder.

kaitj commented on snakebids/core/input_generation.py:
```diff
@@ -233,6 +229,40 @@ def generate_inputs(
+            pybids_database_dir is not None
+            and not Path(pybids_database_dir).is_absolute()
+        ):
+            pybids_database_dir = os.path.join(bids_dir, pybids_database_dir)
```
Fixed in c42049c

tkkuehn commented:
Some good discussion on the interface issue at the dev lunch today, here's an attempt to summarize the main questions we're left with:

- How do we activate the DB feature?
- How do we specify the path to the DB file?
  - Are relative paths allowed? If so, how do we resolve them (i.e. what are they relative to)?
- How do we specify if the database is to be reset?

The answer to each of the questions basically boils down to:

1. Via the config file.
2. Via the call to `generate_inputs` in the Snakefile.
3. Via the command line.

With the current implementation, everything is handled by the config file, and relative paths are resolved relative to the input directory. My preference would be to instead resolve relative paths relative to the output directory, but then `generate_inputs` would need another argument with the output directory, which might end up redundant.

## Potential Solutions

### Config file only, paths relative to input directory

```
# pybids_db:
#   database_dir: '/path/to/db_dir' # Leave blank if you do not wish to use this
#   reset_database: False # Change this to true to update the database
```
```
config.update(
    snakebids.generate_inputs(
        bids_dir=config["bids_dir"],
        pybids_inputs=config["pybids_inputs"],
        pybids_db=config.get("pybids_db"),
        pybids_reset_database=False,
        derivatives=config["derivatives"],
        participant_label=config["participant_label"],
        exclude_participant_label=config["exclude_participant_label"]
    )
)
```
Pros: Already implemented, conceptually simple, minimal changes to `generate_inputs` interface, backwards compatible.
Cons: Might prefer not to default to putting pybids db in the input directory

Variant: Absolute paths only (less flexible but simpler implementation and no questionable default).

### Everything in generate_inputs call

```
config.update(
    snakebids.generate_inputs(
        bids_dir=config["bids_dir"],
        pybids_inputs=config["pybids_inputs"],
        derivatives=config["derivatives"],
        participant_label=config["participant_label"],
        exclude_participant_label=config["exclude_participant_label"]
        pybids_db=config["output_dir"],
    )
)
```
Pros: Easily customizable with full use of the config dict
Cons: Need to go in and edit the workflow any time you want to change anything with the DB.

To be continued on Monday...

tkkuehn commented:
## Potential Solutions Continued

### CLI Only

`run.py {bids_dir} {out_dir} {analysis_level} --pybids_db {db_path} --reset_pybids_db`

Could implement this in `cli.py` as part of the standard parser or add them to the standard `parse_args` mapping in the template config.

Pros: No path ambiguity (assuming `db_path` gets resolved relative to the working directory where `run.py` is called), easy to change DB use from run to run.
Cons: Adds a lot of length to CLI invocations every run if a user always wants to use the pybids DB.

### A combination of the above

For example, could allow the pybids DB location to be defined in the config file (maybe only as an absolute path), but have a CLI argument that overrides it if present. All the pros and cons of whatever solutions are present, plus:

Pros: Flexibility for the user to use whatever works best for their use case.
Cons: Need to make clear how we handle the case where the DB location is defined differently in different places.

tkkuehn commented:
Idea from today's meeting: Combine config (absolute paths only) and CLI (resolve path before adding it to the config dict), with CLI taking precedence if a location is defined on both.

kaitj commented:
I think this is mostly ready for review (thanks @tkkuehn for the help with the CLI).

Only thing missing is testing to see if config is updated from the CLI. Still trying to figure out the best way to do this (probably making use of the mock object).

pvandyken commented on snakebids/tests/test_generate_inputs.py:
```diff
@@ -239,3 +245,70 @@ def test_get_lists_from_bids():
         }
+
+
+def test_db(tmpdir):
```
Could this be broken down into a class of sub-tests? e.g. `test_no_database_when_database_dir_blank`, `test_no_database_when_database_dir_relative`, `test_database_caches_database_dir_absolute`, or something like that?

kaitj replied:
See [cef943a](https://github.com/akhanf/snakebids/pull/133/commits/cef943a2617519fd0e5eddc02a47547c7475803e)

pvandyken commented:
`test_update_config` is intended to test the function called `update_config()` in `app.py`. The best current model for testing such changes is the existing tests in `test_app`.

pvandyken commented:
It does check what args were passed to the `write_config_file` function:
```py
io_mocks["write_config"].assert_called_once_with(
    config_file=Path("/tmp/output/mock/config.yaml"),
    data=expected_config,
    force_overwrite=True,
)
```

Admittedly, `write_config_files` has no test of its own...

pvandyken commented:
Yeah, that's right. More precisely, its `write_config_file` that needs a test

kaitj commented:
> Yeah, that's right. More precisely, its write_config_file that needs a test

Yeah, I mean it probably is worth adding in a test to make sure the args are passed, but I was thinking more of the `write_config_file` test, and that should probably be a separate PR that considers more than just the DB update.

kaitj commented:
Think I figured it out - added the two new values `pybids_db_reset` and `pybids_db_dir` for testing to at least make sure it is properly passed along to the `write_config_file` in both bidsapp and workflow mode.

kaitj commented:
Brought up during the dev lunch today:

Uncomment out the config values, setting `pybids_db_dir` to `None` and adding in a separate flag/value to opt in (with a default suggested path).

tkkuehn commented:
@kaitj This needs a merge from main, then I'll take a last look and merge.

tkkuehn reviewed (approved):
Looks good to me! Will merge now.

pvandyken commented on snakebids/project_template/{{cookiecutter.app_name}}/workflow/Snakefile:
```diff
@@ -10,6 +10,8 @@ config.update(
         bids_dir=config["bids_dir"],
         pybids_inputs=config["pybids_inputs"],
+        pybids_database_dir=config.get("pybids_db_dir"),
+        pybids_reset_database=config["pybids_db_reset"],
```
This needs to be `config.get` or it will fail on the default, commented out state

tkkuehn replied:
Ah in workflow mode this is true, `config["pybids_db_reset"]` gets overwritten from the CLI [here](https://github.com/akhanf/snakebids/pull/133/files#diff-7474454e76eaaba6cbcaa4ebc844258048892849c14d25afa49f0f3644422d2bR193).

kaitj replied:
Updated this to `config.get("pybids_db_reset")`

pvandyken commented on snakebids/tests/test_cli.py:
```diff
@@ -116,3 +116,19 @@ def test_dash_syntax_in_config_cli_args(parser: ArgumentParser, mocker: MockerFi
     assert args.args_dict["arg_using_dash_syntax"][0] == "7890"
+
+
+def test_update_pybids_args(parser: ArgumentParser, mocker: MockerFixture):
```
I think this test isn't doing anything? I'd remove it on the PR, throw it on a different branch if for future reference

kaitj replied:
Removed from the current PR

pvandyken commented on snakebids/tests/test_generate_inputs.py:
```diff
@@ -303,3 +309,80 @@ def test_get_lists_from_bids():
+            pybids_database_dir=self.pybids_db.get("database_dir"),
+            pybids_reset_database=self.pybids_db.get("reset_database"),
+        )
+        assert not self.pybids_db.get("database_dir")
```
Are there any actual checks on the file system here? It looks like only a trivial assertion on the `pybids-db` dict is being made

kaitj replied:
Ahh, you're right - at one point it was checking the file system but not in this case. Will update this assertion.

kaitj replied:
This should be fixed by checking the file path now

pvandyken commented:
Also, could you squash to 1 or two commits before merging?

kaitj commented:
> Also, could you squash to 1 or two commits before merging?

@pvandyken Is there an easy way to do this (a lot of commits happened since the beginning)? Actually just pushed what I thought was squashed history that ended up messing up the git history and had to revert to f9068a5.

pvandyken commented:
I usually just rebase... Something like `git rebase -i HEAD~40` in this case. You can `ctrl+v` block select in vim to change all the `picks` to `squash` at the same time. And make the first one `reword`.

tkkuehn commented:
> > Also, could you squash to 1 or two commits before merging?
> 
> @pvandyken Is there an easy way to do this (a lot of commits happened since the beginning)?

More info on this, for reference: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_squashing

kaitj commented:
Yeah, the problem is coming from the merge from main - that history is in the logs too (probably should have rebased there)

tkkuehn commented:
Another option that's particularly useful if you want to just squash everything into one commit and don't mind losing the commit messages: https://git-scm.com/book/en/v2/Git-Tools-Reset-Demystified#_squashing_2

kaitj commented:
Just managed to squash to a single commit.

Issue was the merge from main rather than a rebase. Got that resolved by just rebasing from main instead.

pvandyken commented:
In the spirit of rebase tips, reflog is a vital tool for recovering from rebase errors. (https://www.atlassian.com/git/tutorials/rewriting-history/git-reflog). It's saved me a few times

kaitj commented:
Alright, after a lot of back and forth between commits, I think finally caught and fixed most of the issues that I introduced from rebasing / squashing. Definitely my least favourite part of git :sweat_smile:

pvandyken commented:
Looks like the test reverted back to a single test for all three cases, rather than 3 separate tests?

kaitj commented:
> Looks like the test reverted back to a single test for all three cases, rather than 3 separate tests?

Should be fixed in latest commit.

tkkuehn reviewed (approved):
Alright, merging!

pvandyken commented:
@kaitj, finally got a chance to use this feature in the wild... it's soooo much faster! Definitely worth all the trouble it took.
